File: totoggle_go/internal/strategy/timewindow.go
package strategy

import (
	"strings"
	"time"

	"github.com/manorfm/toToggles/totoggle_go/internal/toggle"
)

// TimeWindowEvaluator matches a "HH:mm-HH:mm" 24h daily window against the current time. Needs
// no parameter — it reads the clock instead. An overnight window (start > end, e.g.
// "22:00-06:00") wraps past midnight.
type TimeWindowEvaluator struct {
	now func() time.Time
}

// NewTimeWindowEvaluator builds a TimeWindowEvaluator. now, if non-nil, is the clock to read the
// current time from — tests inject a fixed instant instead of depending on wall-clock time. A
// nil now defaults to time.Now.
func NewTimeWindowEvaluator(now func() time.Time) TimeWindowEvaluator {
	if now == nil {
		now = time.Now
	}
	return TimeWindowEvaluator{now: now}
}

func (e TimeWindowEvaluator) Evaluate(rule toggle.ActivationRule, key string, hasKey bool) bool {
	parts := strings.SplitN(rule.Value, "-", 2)
	if len(parts) != 2 {
		return false
	}

	start, err := minutesOfDayFromHHMM(strings.TrimSpace(parts[0]))
	if err != nil {
		return false
	}
	end, err := minutesOfDayFromHHMM(strings.TrimSpace(parts[1]))
	if err != nil {
		return false
	}

	now := minutesOfDay(e.now())
	if start <= end {
		return now >= start && now < end
	}
	// Overnight window, e.g. 22:00-06:00.
	return now >= start || now < end
}

func minutesOfDay(t time.Time) int {
	return t.Hour()*60 + t.Minute()
}

func minutesOfDayFromHHMM(s string) (int, error) {
	parsed, err := time.Parse("15:04", s)
	if err != nil {
		return 0, err
	}
	return minutesOfDay(parsed), nil
}

File: totoggle_node/src/internal/strategy/timewindow.ts
import type { ActivationRule } from "../toggle/rule.js";
import type { Evaluator } from "./strategy.js";

const HH_MM = /^(\d{2}):(\d{2})$/;

function parseHHMM(raw: string): number | null {
  const match = HH_MM.exec(raw.trim());
  if (!match) {
    return null;
  }
  const hours = Number(match[1]);
  const minutes = Number(match[2]);
  if (hours > 23 || minutes > 59) {
    return null;
  }
  return hours * 60 + minutes;
}

/** Minutes since midnight for `date`, read in `timeZone` (undefined = the runtime's own zone) —
 * `Intl.DateTimeFormat` (built into Node/V8, no dependency) resolves the absolute instant `date`
 * represents into that zone's wall-clock time. */
function minutesOfDay(date: Date, timeZone: string | undefined): number {
  const parts = new Intl.DateTimeFormat("en-US", {
    timeZone,
    hour: "2-digit",
    minute: "2-digit",
    hourCycle: "h23",
  }).formatToParts(date);
  const hour = Number(parts.find((p) => p.type === "hour")?.value);
  const minute = Number(parts.find((p) => p.type === "minute")?.value);
  return hour * 60 + minute;
}

/**
 * Matches a "HH:mm-HH:mm" 24h daily window against the current time. Needs no key — it reads the
 * clock instead. An overnight window (start > end, e.g. "22:00-06:00") wraps past midnight.
 */
export class TimeWindowEvaluator implements Evaluator {
  constructor(
    private readonly now: () => Date = () => new Date(),
    private readonly timeZone?: string,
  ) {}

  evaluate(rule: ActivationRule, _key: string | undefined): boolean {
    const dashIndex = rule.value.indexOf("-");
    if (dashIndex === -1) {
      return false;
    }
    const start = parseHHMM(rule.value.slice(0, dashIndex));
    const end = parseHHMM(rule.value.slice(dashIndex + 1));
    if (start === null || end === null) {
      return false;
    }

    const nowMinutes = minutesOfDay(this.now(), this.timeZone);
    if (start <= end) {
      return nowMinutes >= start && nowMinutes < end;
    }
    // Overnight window, e.g. 22:00-06:00.
    return nowMinutes >= start || nowMinutes < end;
  }
}
